text/204-golang-bindings.md

```diff
@@ -0,0 +1,669 @@
+
+    SecurityGroupId() string
+    AllowAllOutbound() bool
+    AddIngressRule(peer: IPeer, connect: Port, description: *string, remoteRule: *boolean)
+    AddEgressRule(peer: IPeer, connect: Port, description: *string, remoteRule: *boolean)
+}
+```
+
+### JSII Datatype Interfaces (`IResource`)
+
+In the JSII, the InterfaceType has a [datatype field](https://github.com/aws/jsii/blob/master/packages/%40jsii/spec/lib/assembly.ts#L879-L888)
+attribute that indicates that the interface only contains properties. While this does corresponds directly to a Go struct, we would likely need to
+generate both a Go interface that contains getter methods that correspond to each property as well as a Go struct that implements that interface. This
+is in order to support subtyping, as the interface is typically what is passed as an argument into other functions, as well as to ensure forward
+compatibility in case the datatype interface eventually extends another. Were it not for these considerations, it would be simpler to simply have a
+single Go struct that corresponds to a datatype interface (see last bullet point in Notes/Concerns).
+
+#### Case 1: Typescript datatype interface (no extensions)
+
+Here, a datatype interface (`HealthCheck`) is converted into a Go interface (`HealthCheck`), where each property becomes a
+getter method (JSII datatype properties always are readonly, so there are no setter methods generated). This Go interface is then implemented by a struct
+(LinuxParameterProps), which would contain the properties from the original typescript interface as non-exported (i.e. lowercase) properties and rely
+on the generated Getter methods for readonly access (as when _used as an argument_ to the constructor):
+
+[Example](https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-ecs/lib/container-definition.ts#L596):
+
+```ts
+export interface HealthCheck {
+  readonly command: string[];
+  readonly interval?: cdk.Duration;
+  readonly retries?: number;
+  readonly startPeriod?: cdk.Duration;
+  readonly timeout?: cdk.Duration;
+}
+
+function renderHealthCheck(hc: HealthCheck): CfnTaskDefinition.HealthCheckProperty {
+  return {
+    command: getHealthCheckCommand(hc),
+    interval: hc.interval != null ? hc.interval.toSeconds() : 30,
+    retries: hc.retries !== undefined ? hc.retries : 3,
+    startPeriod: hc.startPeriod && hc.startPeriod.toSeconds(),
+    timeout: hc.timeout !== undefined ? hc.timeout.toSeconds() : 5,
+  };
+}
+```
+
+Go translation (Run example in the [Go playground](https://play.golang.org/p/_eRLhVmXmp8):
+
+```go
+package ecs
+
+type HealthCheckIface interface{
+   Command() []string
+   Interval() cdk.Duration;
+   Retries() int
+   StartPeriod() cdk.Duration;
+   Timeout() cdk.Duration;
+}
+
+type HealthCheck struct {
+    command []string
+    interval cdk.Duration;
+    retries int
+    startPeriod cdk.Duration;
+    timeout cdk.Duration;
+}
+
+func (h HealthCheck) Command() []string {
+    return h.command
+}
```

Comment:
To truely be immutable, should slice types be copied? What is the depth of the immutability for these types?

Reply:
Kind of an edge case, but asking because the `[]string` returned by `Command()` method in this example, could have elements in the slice reordered, and replaced. A guard against this would be to always create a copy when ever returning lists/maps.

```go
func (h HealthCheck) Command() []string {
     return append([]string{}, h.command...)
}
```